### PR TITLE
add cmake support

### PR DESCRIPTION
**Pull request by mrgreywater**
This adds the CMakeLists.txt that I used to compile this project with CLion on Linux and generate the project files for Visual Studio 2013 and 2015 on Windows.

To build with Visual Studio (assuming cmake ([download](https://cmake.org/download/)) and glfw3 ([download](http://www.glfw.org/download.html)) are properly installed, navigate to the project root directory and execute these commands:
```batch
mkdir vs2015
cd ./vs2015
::in the following commands, change path-to-glfw to your glfw install dir
set GLFW_INCLUDE_DIRS=path-to-glfw\include 
set GLFW_LIBRARIES=path-to-glfw\lib-vc2015\glfw3.lib
cmake .. -G "Visual Studio 14 2015"
```
and open the generated sln project file with Visual Studio (equivalent steps for VS2013)

To build it for linux, make sure to have cmake 3.2 ([ubuntu-repository](https://launchpad.net/~george-edison55/+archive/ubuntu/cmake-3.x)) and glfw3 ([how-to](https://github.com/shurcooL/reusable-commands/blob/ed33ae496f36aaea735a1d183f77e833c92a9f3d/go-gl_glfw3_install.sh#L19-L32)) installed, navigate to the project root directory in your terminal and type:
```bash
mkdir build
cd ./build
cmake .. 
make -j4
```
Other targets were not tested, but should work aswell.

If this PR is merged, I would suggest to add cmake to Travis aswell, since otherwise it may happen quickly that someone edits the gyp or cmake configuration alone without checking the other configuration.

Additionally, in a follow-up PR, we could add a github submodule link to https://github.com/glfw/glfw, and compile against the local glfw source (e.g [like this project](https://github.com/progschj/OpenGL-Examples)), which would make it alot easier to get the project working initally. That way the manual compilation of glfw3 on linux, or setting up environment variables on windows  would not be required.

### CODE REVIEW

**Comment by mrgreywater** [comment]
ref #21

**Comment by jfirebaugh** [comment]
We'll want to switch over to cmake entirely if we go this route, not maintain two build systems in parallel.

cc @kkaefer

**Comment by mrgreywater** [comment]
said follow-up PR with glfw as submodule can be found here: https://github.com/mrgreywater/earcut.hpp/commit/d658b8bfc400d37da95eedcbc9962627f0a03aad

usage would be:
```bash
git submodule init
git submodule update
mkdir build
cd ./build
cmake .. 
make -j4
```

and it would remove the requirement to install glfw manually.

**Comment by mourner** [comment]
Great! @mrgreywater yeah, let's remove the old build system fully in favor of CMake.

**Comment by mourner** [comment]
@mrgreywater BTW, just added you as the repo collaborator. Thank you for all the awesome contributions! Now you can make branches in the upstream repo (and manage issues), but be sure to still submit PRs for any significant changes.

**Comment by mrgreywater** [comment]
@mourner wow, thanks.
I would appreciate if someone else could merge this PR and do the appopriate changes to travis. I am also not quite sure which files can be removed related to gyp.
I guess it's the makefile, the .mason submodule, configure, and all the .gpyi and .gpy files.

**Comment by kkaefer** [comment]
Just a heads up that using CMake 3 on Travis requires adding these sources/packages:

```yaml
addons:
  apt:
    sources: [ 'george-edison55-precise-backports' ]
    packages: [ 'cmake', 'cmake-data' ]
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    # some flags that may be used in the future: -pedantic -Wconversion -Wall
+    if(CMAKE_BUILD_TYPE EQUAL "DEBUG")
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0")
+        # optional: -fsanitize=signed-integer-overflow,undefined
+    else()
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
+        # optional: -march=native (builds with the optimizations available on the build machine (only for local use!))
+    endif()
+else()
+    message(WARNING "Building earcut with unknown compiler.")
```
Comment (kkaefer) [inline CMakeLists.txt]: This message is triggered when compiling with Clang. It means that Clang won't receive any compiler flags.

**Comment by kkaefer** [comment]
As part of this PR, we should also completely remove the existing build system, as we won't be maintaining two systems side-by-side.

File: CMakeLists.txt (hunk 1 above)
Comment (mrgreywater) [inline CMakeLists.txt]: fixed

**Comment by mourner** [comment]
@kkaefer want to take a stab at this? See https://github.com/mapbox/earcut.hpp/pull/34#issuecomment-184891113

**Comment by mrgreywater** [comment]
@kkaefer I gave you access to https://github.com/mrgreywater/earcut.hpp/tree/cmake in case you want to make changes to this PR yourself. Otherwise I may give try to remove gyp and make the changes to Travis myself later, even though I cannot tell how difficult it will get as I don't really know how travis/gyp works.

**Comment by mrgreywater** [comment]
@kkaefer @jfirebaugh @mourner looks like it works now. Would be nice if one of you could check if I haven't missed anything/ deleted to much related to gyp.

**Comment by mourner** [comment]
Oh wow! Isn't this just lovely? 
image

**Pull request merged**